README.md
# FeatureServer

## DEPRECATED - now migrated to a package in the [Koop monorepo](https://github.com/koopjs/koop/).  Update any `npm` usage to `@koopjs/featureserver`.

[![No Maintenance Intended](http://unmaintained.tech/badge.svg)](http://unmaintained.tech/)
[![npm][npm-image]][npm-url]
[![travis][travis-image]][travis-url]
[![Greenkeeper badge][greenkeeper-image]][greenkeeper-url]

*An open source implementation of the GeoServices specification*

## Usage

This is meant to be used as a plugin to Express

### Example server
```js
const express = require('express')
const app = express() // set up a basic express server
const FeatureServer = require('featureserver')
const cache = require('cache')

// We only need one handler because FeatureServer.route is going to do all the work
const handler = (req, res) => {
  cache.get(/* some geojson */, (err, data) => {
    if (err) res.status(500).json({error: err.message})
    else FeatureServer.route(req, res, data)
  })
}

// Sets up all of the handled routes to support `GET` and `POST`
const routes = ['/FeatureServer', '/FeatureServer/layers', '/FeatureServer/:layer', '/FeatureServer/:layer/:method']

routes.forEach(route => {
  app.route(route)
  .get(handler)
  .post(handler)
})
```

### Setting defaults at runtime
FeatureServer allows several defaults to be set at runtime via Express's `app.locals` method.  Specifically, you will need to set:

```js
app.locals.config = {
  featureServer: {
    // define default here
  }
}
```

If you are using FeatureServer as part of a Koop instance, the equivalent of Express's `app.locals` is `koop.server.locals`.

The follow properties can be set at runtime with the noted method:

```js
app.locals.config = {
  featureServer: {
    currentVersion: 11.01, // defaults to 10.51
    fullVersion: '11.0.1', // defaults to '10.5.1'
    serviceDescription: 'default service description',
    description: 'default layer description'
  }
}
```

## API
* [FeatureServer.route](#FeatureServer.route)
* [FeatureServer.query](#FeatureServer.query)
* [FeatureServer.serverInfo](#FeatureServer.serverInfo)
* [FeatureServer.layerInfo](#FeatureServer.layerInfo)
* [FeatureServer.layers](#FeatureServer.layers)
* [FeatureServer.generateRenderer](#FeatureServer.generateRenderer)
* [FeatureServer.authenticate](#FeatureServer.authenticate)
* [FeatureServer.error.authorize](#FeatureServer.error.authorize)
* [FeatureServer.authenticate](#FeatureServer.error.authenticate)
* [FeatureServer.queryRelatedRecords](#FeatureServer.queryRelatedRecords)

### FeatureServer.route
Pass in an `incoming request object`, an `outgoing response object`, a `geojson` object, and `options` and this function will route and return a geoservices compliant response

- Supports: '/FeatureServer', '/FeatureServer/layers', '/FeatureServer/:layer', '/FeatureServer/:layer/:method'
	- _Note_: only `query`, `info`, and `generateRenderer` are supported methods at this time.

```js
FeatureServer.route(req, res, data, options)
```

- **data** is either a geojson object extended with some additional properties or an object with a layers property which an array of extended geojson objects. These properties are optional and can be used to provide more specific metadata or to shortcut the built in filtering mechanism.

e.g.

```js
{
  type: 'FeatureCollection' // Static
  features: Array, // GeoJSON features
  statistics: Object, // pass statistics to an outStatistics request to or else they will be calculated from geojson features passed in
  metadata: {
    id: number, // The unique layer id.  If supplied for one layer, you should supply for all layers to avoid multiple layers having the same id.
    name: String, // The name of the layer
    description: String, // The description of the layer
    copyrightText: String, // The copyright text (layer attribution text)
    extent: Array, // valid extent array e.g. [[180,90],[-180,-90]]
    displayField: String, // The display field to be used by a client
    geometryType: String // REQUIRED if no features are returned with this object Point || MultiPoint || LineString || MultiLineString || Polygon || MultiPolygon
    idField: String, // unique identifier field,
    maxRecordCount: Number, // the maximum number of features a provider can return at once
    limitExceeded: Boolean, // whether or not the server has limited the features returned
    timeInfo: Object, // describes the time extent and capabilities of the layer,
    transform: Object, // describes a quantization transformation
    renderer: Object, // provider can over-ride default symbology of FeatureServer output with a renderer object. See https://developers.arcgis.com/web-map-specification/objects/simpleRenderer, for object specification.
    defaultVisibility: boolean, // The default visibility of this layer
    minScale: number, // The minScale value for this layer
    maxScale: number, // The maxScale value for this layer
    fields: [
     { // Subkeys are optional
       name: String,
       type: String, // 'Date' || 'Double' || 'Integer' || 'String'
       alias: String, // how should clients display this field name,
     }
    ]
  },
  capabilities: {
    quantization: Boolean // True if the provider supports quantization
  },
  filtersApplied: {
    all: Boolean // true if all post processing should be skipped
    geometry: Boolean, // true if a geometric filter has already been applied to the data
    where: Boolean, // true if a sql-like where filter has already been applied to the data
    offset: Boolean // true if the result offset has already been applied to the data,
    limit: Boolean // true if the result count has already been limited,
    projection: Boolean // true if the result data has already been projected
  }
  count: Number // pass count if the number of features in a query has been pre-calculated
}
```

or

```js
{
  layers: [
    {
      type: 'FeatureCollection'
      ...
    },
    {
      type: 'FeatureCollection'
      ...
    }
]
```
- **options** is an object that dictates method actions. See `FeatureServer.query` and `FeatureServer.generateRenderer` for more details.

### FeatureServer.query
Pass in `geojson` and `options` (a valid [geoservices query object](https://geoservices.github.io/query.html)), and the function will perform the query and return a valid geoservices query object. The in addition to input `statistics: {}`, following is an example of _all_ query `options` that can be passed into the query route: '/FeatureServer/:layer/query'

e.g.

```js
const options = {
  where: `1=1`,
  objectIds: '1,2,3',
  geometry: {
    xmin: -110, ymin: 30, xmax: -106, ymax: 50,
    spatialReference: { wkid: 4326 },
  },
  geometryType: 'esriGeometryEnvelope',
  spatialRel: 'esriSpatialRelContains',
  outFields: '*',
  returnGeometry: true,
  outSR: 102100, // output spatial reference
  returnIdsOnly: true,
  returnCountOnly: true,
  orderByFields: 'Full/Part_COUNT DESC',
  groupByFieldsForStatistics: 'Full/Part',
  outStatistics: {
  	statisticType: 'count',
  	onStatisticField: '<field>',
  	outStatisticFieldName: 'name'
  },
  returnDistinctValues: true,
  resultOffset: 0,
  resultRecordCount: 0,
  f: 'pjson'
}

FeatureServer.query(geojson, options)
```

### FeatureServer.serverInfo
Generate version `10.51` Geoservices server info

```js
const server = {
  description: String // Describes the collection of layers below,
  copyrightText: String // Optional copyright text
  maxRecordCount: Number // the maximum number of features a provider can return at once,
  hasStaticData: Boolean // whether or not the server contains any data that is not changing
  hasAttachments: Boolean // whether or not the server contains any attachments for this layer
  layers: [{ // A collection of all the layers managed by the server
    type: 'FeatureCollection',
    metadata: {
      id: number, // The unique layer id.  If supplied for one layer, you should supply for all layers to avoid multiple layers having the same id.
      name: String, // The name of the layer
      description: String, // The description of the layer
      extent: Array, // valid extent array e.g. [[180,90],[-180,-90]]
      displayField: String, // The display field to be used by a client
      idField: String, // unique identifier field,
      geometryType: String, // REQUIRED if no features are returned with this object Point || MultiPoint || LineString || MultiLineString || Polygon || MultiPolygon
      maxRecordCount: Number, // the maximum number of features a provider can return at once
      limitExceeded: Boolean, // whether or not the server has limited the features returned
      timeInfo: Object, // describes the time extent and capabilities of the layer
      renderer: Object, // provider can over-ride default symbology of FeatureServer output with a renderer object. See https://developers.arcgis.com/web-map-specification/objects/simpleRenderer, for object specification.
      defaultVisibility: boolean, // The default visibility of this layer
      minScale: number, // The minScale value for this layer
      maxScale: number // The maxScale value for this layer
    }
    features: [// If all the metadata provided above is provided features are optional.
      {
        type: 'Feature',
        geometry: {
          type: 'Point',
          coordinates: [125.6, 10.1]
        },
        properties: {
          name: 'Dinagat Islands'
        }
      }]
    }
  }],
  tables: [{ // A collection of all the tables managed by the server
    type: 'FeatureCollection',
    metadata: {
      // see layer metadata
    }
  }],
  relationships: [{ // A collection of all relationships manged by the server
    id: number, // The unique relationship id.
    name: String, // The name of the relationship
  }]
}

FeatureServer.serverInfo(server)
```

### FeatureServer.layerInfo
Generate version `10.51` Geoservices information about a single layer
```js
FeatureServer.layerInfo(geojson, options)
```

Note that the layer info is modified with properties `metadata` and `capabilites` found at the top-level of the GeoJSON object.

|GeoJSON property| Layer info result|
|---|---|
|`metadata.id`| overrides default|
|`metadata.name`| overrides default|
|`metadata.description`| overrides default |
|`metadata.geometryType`| overrides value determined from data |
|`metadata.extent`| overrides value determined from data |
|`metadata.timeInfo`| overrides default |
|`metadata.maxRecordCount`| overrides default (2000) |
|`metadata.displayField`| overrides default (`OBJECTID`) |
|`metadata.objectIdField`| overrides default  (`OBJECTID`) |
|`metadata.hasStaticData`| overrides default (`false`) |
|`metadata.hasAttachments`| overrides default (`false`) |
|`metadata.renderer`| overrides default |
|`metadata.defaultVisibility`| overrides default |
|`metadata.minScale`| overrides default |
|`metadata.maxScale`| overrides default |
|`metadata.relationships` | overrides default |
|`capabilities.extract`|  when set to `true`, `Extract` added to `capabilites` (e.g., `capabilities: "Query,Extract"`) |
|`capabilities.quantization`| when set to `true`, `supportsCoordinatesQuantization: true` |

##### metadata.relationships
This defined the server managed relationships for the layer

e.g.

```js
const metadata = {
  //...
  relationships: [{ // A collection of all relationships manged by the server
    id: number, // The unique relationship id.
    name: String, // The name of the relationship
    relatedTableId: number, // Id of the layer/table related records are found
    cardinality: String, // esriRelCardinalityOneToMany | esriRelCardinalityManyToMany
    role: String, // esriRelRoleOrigin | esriRelRoleDestination
    keyField: String, // key field name in the related Table 
    composite: Boolean // likely to false
  }]
  //...
}
```

### FeatureServer.layers
Generate version `10.51` Geoservices information about one or many layers

Can pass a single geojson object or an array of geojson objects
```js
FeatureServer.layers(geojson, options)
```

### FeatureServer.generateRenderer
Pass in `geojson` and `options`, and the function will return a valid generateRenderer object. Two `classificationDef` classification types are supported, _classBreaksDef_ and _uniqueValueDef_.

_classBreaksDef_ is used to classify numeric data based on a number of breaks and a statistical method. Features can also be normalized before being classified. _uniqueValueDef_ is used to classify data based on a unique field(s). If classification breaks are not supplied through in `statistics`, they will be generated using `classificationDef` options. The output is a _generateRenderer_ object.

##### classBreaksDef
In addition to class breaks as input `statistics: []`, the following is an example of _all_ classBreaksDef `options` that can be passed into the generateRenderer route: '/FeatureServer/:layer/generateRenderer'

e.g.

```js
const options = {
 *'classificationDef': {
   *'type': 'classBreaksDef',
   *'classificationField': '<field1>',
   *'classificationMethod': 'esriClassifyEqualInterval' | 'esriClassifyNaturalBreaks' | 'esriClassifyQuantile' | 'esriClassifyStandardDeviation',
   *'breakCount': 9,
    'normalizationType': 'esriNormalizeByField' | 'esriNormalizeByLog' | 'esriNormalizeByPercentOfTotal',
    'normalizationField': '<field2>' // mandatory if 'normalizationType' === 'esriNormalizeByField'
    'baseSymbol': {
      'type': 'esriSMS',
      'style': 'esriSMSCircle',
      'width': 2
    },
    'colorRamp': {
      'type': 'algorithmic',
      'fromColor': [115,76,0,255],
      'toColor': [255,25,86,255],
      'algorithm': 'esriHSVAlgorithm'
    }
  },
  'where': '<field2> > 39'
}

FeatureServer.generateRender(geojson, options)

*required
```

Output:

```js
{
  type: 'classBreaks',
  field: '<field1>',
  classificationMethod: 'esriClassifyEqualInterval',
  minValue: 0,
  classBreakInfos: [
    {
      classMinValue: 0,
      classMaxValue: 5,
      label: '0-5',
      description: '',
      symbol: {
        type: 'esriSMS',
        style: 'esriSMSCircle',
        width: 2,
        color: [115, 76, 0]
      }
    },
    {
      classMinValue: 6,
      classMaxValue: 11,
      label: '6-11',
      description: '',
      symbol: {
        type: 'esriSMS',
        style: 'esriSMSCircle',
        width: 2,
        color: [156, 67, 0]
      }
    },
    ...
  ]
}
```

##### uniqueValueDef
The following is an example of _all_ uniqueValueDef `options` that can be passed into the generateRenderer route: '/FeatureServer/:layer/generateRenderer'

e.g.

```js
const options = {
 *'classificationDef': {
   *'type': 'uniqueValueDef',
   *'uniqueValueFields': ['Genus', '<field2>', '<field3>'],
   *'fieldDelimiter': ', '
    'baseSymbol': {
      'type': 'esriSMS',
      'style': 'esriSMSCircle',
      'width': 2
    },
    'colorRamp': {
      'type': 'algorithmic',
      'fromColor': [115,76,0,255],
      'toColor': [255,25,86,255],
      'algorithm': 'esriHSVAlgorithm'
    }
  },
  'where': 'latitude > 39'
}

FeatureServer.generateRender(geojson, options)

*required
```

Output:

```js
{
  type: 'uniqueValue',
  field1: 'Genus',
  field2: '',
  field3: '',
  fieldDelimiter: ', ',
  defaultSymbol: {},
  defaultLabel: '',
  uniqueValueInfos: [
    {
      value: 'MAGNOLIA',
      count: 5908,
      label: 'MAGNOLIA',
      description: '',
      symbol: {
        type: 'esriSMS',
        style: 'esriSMSCircle',
        width: 2,
        color: [115, 76, 0]
  	   }
    },
    {
      value: 'QUERCUS',
      count: 12105,
      label: 'QUERCUS',
      description: '',
      symbol: {
        type: 'esriSMS',
        style: 'esriSMSCircle',
        width: 2,
        color: [116, 76, 0]
  	  }
   },
   ...
  ]
```

### FeatureServer.authenticate
Pass in an outgoing response object and an authentication success object and this function will route and return a formatted authentication success response.

    FeatureServer.authenticate(res, auth, ssl = false)

* `auth` is the result of a successful authentication attempt that returns a token and expiration time
* `ssl` is a boolean flag indicating if token should always be passed back via HTTPS. Defaults to `false`

e.g.,
    
    const auth = {
      "token":"elS39KU4bMmZQgMXDuswgA14vavIp4mfpiqcWSr0qM6q4dFguTnnHddWqbpK5Mc3HsCN8XghlwawUUYApOOcxKNyg_9WqTofChJXxxD058_rL1HZkM5PDhUOh9YYQn1K",
      "expires":1524508236322
    }

    FeatureServer.authenticate(res, auth)

    {
      "token":"elS39KU4bMmZQgMXDuswgA14vavIp4mfpiqcWSr0qM6q4dFguTnnHddWqbpK5Mc3HsCN8XghlwawUUYApOOcxKNyg_9WqTofChJXxxD058_rL1HZkM5PDhUOh9YYQn1K",
      "expires":1524508236322,
      ssl: false
    }

### FeatureServer.error.authorize
Pass in an outgoing response object and this function will route and return a formattted authorization error.

    FeatureServer.error.authorize(res)

    {
      "error": {
        "code": 499,
        "message": "Token Required",
        "details": []
      }
    }

### FeatureServer.error.authenticate
Pass in an outgoing response object and this function will route and return a formatted authentication error.

    FeatureServer.error.authenticate(res)
    
    {
      "error": {
        "code": 400,
        "message": "Unable to generate token.",
        "details": ["Invalid username or password."]
      }
    }


[npm-image]: https://img.shields.io/npm/v/featureserver.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/featureserver
[travis-image]: https://img.shields.io/travis/koopjs/FeatureServer.svg?style=flat-square
[travis-url]: https://travis-ci.org/koopjs/FeatureServer
[greenkeeper-image]: https://badges.greenkeeper.io/koopjs/FeatureServer.svg
[greenkeeper-url]: https://greenkeeper.io/

## Unreleased

### FeatureServer.queryRelatedRecords
Pass in `geojson` and `options`, and the function will return a valid queryRelatedRecords object. Required attributes within `options` are `objectIds` and `relationshipId`.

The `geojson` should be in the special FeatureCollection of FeatureCollections format to show the relationship between requested Features within the layer/table and the referenced relatinoship's features.
e.g. 

```js
  const geojson = {
    "type": "FeatureCollection",
    "features": [ // Array of FeatureCollections by objectId with the related records as features
      {
        "type": "FeatureCollection",
        "properties": {
          "OBJECTID": 37
        },
        "features": [
          {
            "type": "Feature",
            "geometry": {...},
            "properties": {...}
          }
        ]
      } 
    ]
  }
  
  const options = {
    objectIds: "37, 462", // comma separated string of object ids within the layer to get related records 
    relationshipId: 4, // relationship Id of the server manged relationship of the layer, see FeatureServer.layerInfo
  }

FeatureServer.queryRelatedRecords(geojson, options)
```

Output:

```js
{
  "geometryType": "esriGeometryPolygon",
  "spatialReference": {
    "wkid": 4267
  },
  "fields": [
    {
      "name": "OBJECTID", 
      "type": "esriFieldTypeOID", 
      "alias": "OBJECTID"
    }, 
    {
      "name": "FIELD1", 
      "type": "esriFieldTypeString", 
      "alias": "FIELD1", 
      "length": 25
    }
  ],
  "relatedRecordGroups": [
    {
      "objectId": 37,
      "relatedRecords": [
        {
          "attributes": {
            "OBJECTID": 5540,
            "FIELD1": "1000147595"
          },
          "geometry": {...}
        }
      ]
    }
  ]
}
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [5.0.0] - 02-2-2023
### Added
- Add support for returnDistinctValues

### Changed
- No more support for Node 10, 12

## [4.1.1] - 11-04-2022
### Fixed
- Allow field definitions to set editable and nullable properties

## [4.1.0] - 09-15-2022
### Added
- Set hasZ from metadata setting when present

## [4.0.2] - 09-14-2022
### Fixed
- geojson validation security vulnerablity
- fix ids in layers metadata
- fix params passed to QueryFields.create


## [4.0.1] - 08-05-2022
### Changed
- bump winnow version

## [4.0.0] - 06-14-2022
### Added
- override default currentVersion and fullVersion via app.locals.config.featureServer
- override default service-description and layer-description via app.locals.config.featureServer

### Changed
- function signature on the restInfo method
## [3.5.1] - 06-01-2022
### Changed
- bump winnow version
- refactor "fields" array generation

## [3.5.0] - 04-25-2022
### Added
- Add support for overrided default value for copyrightText

## [3.4.0] - 03-28-2022
### Added
- Add support for overriding default value for supportsPagination

## [3.3.4] - 03-25-2022
### Changed
- Refactor design pattern for layer metadata

## [3.3.3] - 03-23-2022
### Fixed
- Server info handler was not properly assigned


## [3.3.2] - 03-11-2022
### Fixed
- Ensure `outSR ` is used when `returnExtentOnly=true` and extent not precalculated

### Changed
- Refactor the query handler

## [3.3.1] - 02-24-2022
### Fixed
- support for `returnExtentOnly`

## [3.3.0] - 02-16-2022
### Added
* support for queryRelatedRecords

## [3.2.1] - 01-20-2022
### Fixed
* templates pulled from metadata for JSAPI 4.x support
* OBJECTID not forced to Integer if it is a true string (GUID)

## [3.2.0] - 11-30-2021
### Added
* Support for multipoint

### Fixed
* Table ID numbering

## [3.1.0] - 09-28-2021
### Added
* Ability to specify the editable property for each field via the metadata object

## [3.0.1] - 09-01-2021
### Changed
* Added ability to specify capabilities the feature service would like to handle

## [3.0.0] - 02-18-2021
### Changed
* End support for Node < 10; remove buble compile step

### Fixed
* Allow other properties in spatialReference object

## [2.25.2] - 12-18-2020
### Fixed
* Properly set spatial reference values in service, layer, and query info when source data is not WGS84

## [2.25.1] - 12-14-2020
### Changed
* Bumped Winnow version for patch

## [2.25.0] - 12-14-2020
### Changed
* Bumped Winnow version - better support for input/output CRSs

## [2.24.1] - 09-10-2020
### Changed
* Bumped Winnow version

## [2.24.0] - 09-08-2020
### Changed
* Bumped Winnow version

## [2.23.0] - 05-13-2020
### Added
* Added Multi versions of all geojson shapes so they will be recognized and translated to esri geometries

## [2.22.1] - 04-14-2020
### Changed
* Bumped Winnow version

## [2.22.0] - 01-13-2020
### Added
* ability to override id, defaultVisibility, minScale, and maxScale in layer info via metadata object

## [2.21.2] - 10-31-2019
### Fixed
* Ensure ObjectID field is moved to being of response fields array
* Pass domain and nullable properties into response if provided

## [2.21.1] - 10-04-2019
### Fixed
* Correct packaging

## [2.21.0] - 09-27-2019
### Added
* Support for provider to set renderer; geojson.metadata.render

## [2.20.0] - 07-02-2019
### Added
* Support for `returnExtentOnly`

## [2.19.0] - 06-27-2019
### Added
* Support for info routes with trailing slashes

## [2.18.0] - 05-29-2019
### Added
* Support for `/FeatureServer/info` requests

## [2.17.2] - 05-22-2019
### Fixed
* `returnIdsOnly=true` query requests has `objectIdFieldName` property rather than `objectIdField`

## [2.17.1] - 04-30-2019
### Added
* Colorized warnings

### Changed
* Condensed GeoJSON warnings
* Provide a more descriptive unassigned `idField` warning

## [2.17.0] - 04-10-2019
### Added
* Additional layer-info route handling for `FeatureServer/:layerId/info`

## [2.16.4] - 12-27-2018
### Fixed
* Before warning of discrepancies between metadata `fields` array and feature properties, compare name and alias to feature property keys.
* Remove `exceededTransferLimit` check specific to `maxRecordCount` which can break paging functionality of clients

## [2.16.3] - 11-28-2018
### Changed
* Removed workaround for adding and `OBJECTID` field even when metadata `idField` is set

## [2.16.2] - 10-19-2018
### Changed
* Bump Winnow to 1.16.4

### Fixed
* Add `resultRecordCount` to the `limit` setting hierarchy

## [2.16.1] - 10-03-2018
### Fixed
* package not pointed to `dist/index.js`

## [2.16.0] - 09-17-2018
### Added
* Support addition of "Extract" to layer info capabilities if provider's geojson includes `{ capabilities: { extract: true } }`

## [2.15.2] - 09-10-2018
### Changed
* Supress various warnings when `NODE_ENV !== 'production'` or `KOOP_WARNINGS !== 'suppress'`

## [2.15.1] - 06-06-2018
### Added
* Add warnings for invalid GeoJSON when `NODE_ENV !== 'production'`

## [2.15.0] - 05-30-2018
### Fixed
* Authentication and authorization error messages now properly handled if request has `callback` query parameter

### Added
* Helper function to handle all responses, with or without `callback` query parameter
* `fullVersion: "10.5.1"` to info responses
* 404 handling of urls that don't include `/FeatureServer`

## [2.14.1] - 05-22-2018
### Fixed
* Version bump due to NPM's inability to resolve last publish

## [2.14.0] - 05-22-2018
### Added
* Add response handlers for authentication success and authentication/authorization errors

## [2.13.0] - 05-10-2018
### Added
* Add parameter to the `rest/info` handler and use the argument to supplement/override of the templated JSON response

## [2.12.1] - 05-03-2018
### Fixed
* Bump version, 2.12.0 was npm published without a compile

## [2.12.0] - 05-03-2018
### Added
* support the `/rest/info` route

lib/query.js
const Winnow = require('winnow')
const esriExtent = require('esri-extent')
const { getCollectionCrs } = require('./helpers')
const { renderFeatures, renderStatistics, renderStats } = require('./templates')
const Utils = require('./utils')
const _ = require('lodash')
const chalk = require('chalk')
const getGeometryTypeFromGeojson = require('./helpers/get-geometry-type-from-geojson')

module.exports = query

/**
 * processes params based on query params
 *
 * @param {object} geojson
 * @param {object} params
 */
function query (geojson, params = {}) {
  if (geojson.statistics) {
    return renderStats(geojson)
  }

  const {
    features,
    filtersApplied: { all: skipFiltering } = {},
    metadata: { idField } = {}
  } = geojson

  const data = skipFiltering || !features ? geojson : postProcessData(geojson, params)

  // Warnings
  if (shouldLogWarnings()) {
    logWarnings(data, modifyParams)
  }

  if (params.f === 'geojson') {
    return {
      type: 'FeatureCollection',
      features: data.features
    }
  }

  return renderGeoservicesResponse(data, {
    ...params,
    idField,
    attributeSample: _.get(geojson, 'features[0].properties'),
    geometryType: getGeometryTypeFromGeojson(geojson)
  })
}

function postProcessData (inputData, params) {
  const postProcessingParams = modifyParams(params, inputData)
  return Winnow.query(inputData, postProcessingParams)
}

function modifyParams (params, data) {
  const { f: format, returnExtentOnly, inputCrs, sourceSR } = params
  const { filtersApplied } = data
  const modifedParams = filtersApplied ? modifyParamsForProcessingAlreadyApplied(params, filtersApplied) : params

  if (shouldTransformToEsriJson(format, returnExtentOnly)) {
    modifedParams.toEsri = true
  }

  modifedParams.inputCrs = getInputCrs(data, { inputCrs, sourceSR })

  return modifedParams
}

function modifyParamsForProcessingAlreadyApplied (params, alreadyApplied) {
  const paramsToOmit = _.chain(alreadyApplied).entries()
    .map(([key, value]) => {
      if (key === 'projection') {
        return ['outSR', key]
      }

      if (key === 'offset') {
        return ['resultOffset', key]
      }

      if (key === 'limit') {
        return ['resultRecordCount', key]
      }

      return key
    })
    .flatten()
    .value()

  return _.omit(params, paramsToOmit)
}

function shouldTransformToEsriJson (requestedFormat, returnExtentOnly) {
  return requestedFormat !== 'geojson' && !returnExtentOnly
}

function getInputCrs (data, { inputCrs, sourceSR }) {
  return inputCrs || sourceSR || _.get(data, 'metadata.crs') || getCollectionCrs(data) || 4326
}

function shouldLogWarnings () {
  return process.env.NODE_ENV !== 'production' && process.env.KOOP_WARNINGS !== 'suppress'
}

function logWarnings (geojson, { toEsri }) {
  const { metadata = {}, features } = geojson
  // ArcGIS client warnings
  if (toEsri && !metadata.idField) {
    console.warn(chalk.yellow('WARNING: requested provider has no "idField" assignment. You will get the most reliable behavior from ArcGIS clients if the provider assigns the "idField" to a property that is an unchanging 32-bit integer. Koop will create an OBJECTID field in the absence of an "idField" assignment.'))
  }

  if (toEsri && hasMixedCaseObjectIdKey(metadata.idField)) {
    console.warn(chalk.yellow('WARNING: requested provider\'s "idField" is a mixed-case version of "OBJECTID". This can cause errors in ArcGIS clients.'))
  }

  // Compare provider metadata fields to feature properties
  // TODO: refactor
  if (metadata.fields && _.has(features, '[0].properties')) {
    warnOnMetadataFieldDiscrepencies(geojson.metadata.fields, geojson.features[0].properties)
  }
}

function hasMixedCaseObjectIdKey (idField = '') {
  return idField.toLowerCase() === 'objectid' && idField !== 'OBJECTID'
}

/**
 * Compare fields generated from metadata to properties of a data feature.
 * Warn if differences discovered
 * @param {*} metadataFields
 * @param {*} properties
 */
function warnOnMetadataFieldDiscrepencies (metadataFields, featureProperties) {
  // build a comparison collection from the data samples properties
  const featureFields = Object.keys(featureProperties).map(name => {
    return { name, type: Utils.detectType(featureProperties[name]) }
  })

  // compare metadata to feature properties; identifies fields defined in metadata that are not found in feature properties
  // or that have a metadata type definition inconsistent with feature property's value
  metadataFields.filter(field => {
    // look for a defined field in the features properties
    const featureField = _.find(featureFields, ['name', field.name]) || _.find(featureFields, ['name', field.alias])
    if (!featureField || (field.type !== featureField.type && !(field.type === 'Date' && featureField.type === 'Integer') && !(field.type === 'Double' && featureField.type === 'Integer'))) {
      console.warn(chalk.yellow(`WARNING: requested provider's metadata field "${field.name} (${field.type})" not found in feature properties)`))
    }
  })

  // compare feature properties to metadata fields; identifies fields found on feature that are not defined in metadata field array
  featureFields.filter(field => {
    const noNameMatch = _.find(metadataFields, ['name', field.name])
    const noAliasMatch = _.find(metadataFields, ['alias', field.name])

    // Exclude warnings on feature fields named OBJECTID because OBJECTID may have been added by winnow in which case it should not be in the metadata fields array
    if (!(noNameMatch || noAliasMatch) && field.name !== 'OBJECTID') {
      console.warn(chalk.yellow(`WARNING: requested provider's features have property "${field.name} (${field.type})" that was not defined in metadata fields array)`))
    }
  })
}

/**
 * Format the queried data according to request parameters
 * @param {object} data - full dataset
 * @param {object} data - subset of data with query applied
 * @param {object} params
 */
function renderGeoservicesResponse (data, params = {}) {
  const {
    objectIds,
    outStatistics,
    returnCountOnly,
    returnExtentOnly,
    returnIdsOnly,
    outSR
  } = params

  if (shouldFilterByObjectIds(objectIds, outStatistics)) {
    data.features = filterByObjectIds(data, objectIds)
  }

  if (returnCountOnly && returnExtentOnly) {
    return {
      count: getCount(data),
      extent: getExtent(data)
    }
  }

  if (returnCountOnly) {
    return {
      count: getCount(data)
    }
  }

  if (returnExtentOnly) {
    return {
      extent: getExtent(data, outSR)
    }
  }

  if (returnIdsOnly) {
    return renderIdsOnly(data)
  }

  if (outStatistics) {
    return queryStatistics(data, params)
  }

  return renderFeatures(data, params)
}

function shouldFilterByObjectIds (objectIds, outStatistics) {
  // request for objectIds ignored if out-statistics option is also requested
  return objectIds && !outStatistics
}

function filterByObjectIds (data, objectIds) {
  const idField = _.get(data, 'metadata.idField') || 'OBJECTID'
  const requestedObjectIds = normalizeObjectIds(objectIds)

  return data.features.filter(({ attributes }) => {
    return requestedObjectIds.includes(attributes[idField])
  })
}

function normalizeObjectIds (objectIds) {
  let ids
  if (Array.isArray(objectIds)) {
    ids = objectIds
  } else if (typeof objectIds === 'string') {
    ids = objectIds.split(',')
  } else if (typeof objectIds === 'number') {
    ids = [objectIds]
  } else {
    const error = new Error('Invalid "objectIds" parameter.')
    error.code = 400
    throw error
  }

  return ids.map(i => {
    if (isNaN(i)) {
      return i
    }

    return parseInt(i)
  })
}

function getCount (data) {
  return data.count !== undefined ? data.count : _.get(data, 'features.length', 0)
}

/**
 * Format a response for an ids-only request
 * @param {object} data
 */
function renderIdsOnly (data) {
  const oidField = _.get(data, 'metadata.idField') || 'OBJECTID'
  const response = { objectIdFieldName: oidField, objectIds: [] }
  response.objectIds = data.features.map(f => { return f.attributes[oidField] })
  return response
}

function queryStatistics (data, params) {
  // This little dance is in place because the stat response from Winnow is different
  // TODO make winnow come out as expected
  // or move this into templates.render
  const statResponse = {
    type: 'FeatureCollection',
    features: []
  }
  const features = Array.isArray(data) ? data : [data]
  statResponse.features = features.map(row => {
    return { attributes: row }
  })
  return renderStatistics(statResponse, params)
}

/**
 * Get an extent object for passed GeoJSON
 * @param {object} geojson
 * @param {*} outSR Esri spatial reference object, or WKID integer
 */
function getExtent (geojson, outSR) {
  if (geojson.extent) {
    return geojson.extent
  }

  // Calculate extent from features
  const extent = esriExtent(geojson)

  if (!outSR) return extent

  // Esri extent assumes WGS84, but the data passed in may have been transformed
  // to a different coordinate system by winnow. Math should be the same for the
  // output spatial references we support, but we need to alter the spatial reference
  // property to reflect the requested outSR

  // when outSR submitted as wkt
  if (outSR.wkt) {
    extent.spatialReference = {
      wkt: outSR.wkt
    }
    return extent
  }

  // When submitted as a WKID
  const wkid = outSR.latestWkid || outSR.wkid || outSR
  if (Number.isInteger(wkid)) {
    extent.spatialReference = { wkid }
    return extent
  }

  return extent
}
